// @progress/kendo-angular-grid/dist/npm/dragdrop/draggable-column.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var Observable_1 = require("rxjs/Observable");
var Subscription_1 = require("rxjs/Subscription");
require("rxjs/add/operator/delay");
require("rxjs/add/operator/do");
require("rxjs/add/operator/filter");
require("rxjs/add/operator/map");
require("rxjs/add/operator/switchMap");
require("rxjs/add/operator/switchMapTo");
require("rxjs/add/operator/takeUntil");
var draggable_directive_1 = require("../common/draggable.directive");
var drag_and_drop_service_1 = require("./drag-and-drop.service");
var drag_hint_service_1 = require("./drag-hint.service");
var drop_cue_service_1 = require("./drop-cue.service");
var utils_1 = require("../utils");
var navigation_service_1 = require("../navigation/navigation.service");
/**
 * @hidden
 */
var stopPropagation = function (event) {
    event.stopPropagation();
    event.preventDefault();
};
/**
 * @hidden
 */
var preventOnDblClick = function (release) { return function (mouseDown) {
    return Observable_1.Observable.of(mouseDown).delay(150).takeUntil(release);
}; };
var hasClass = function (className) { return function (el) { return new RegExp("(^| )" + className + "( |$)").test(el.className); }; };
var isDeleteButton = utils_1.or(hasClass("k-i-group-delete"), hasClass("k-button-icon"));
var isSortIcon = utils_1.or(hasClass("k-i-sort-asc-sm"), hasClass("k-i-sort-desc-sm"));
var skipButtons = utils_1.and(utils_1.not(isDeleteButton), utils_1.not(isSortIcon));
var elementUnderCursor = function (_a) {
    var clientX = _a.clientX, clientY = _a.clientY;
    return document.elementFromPoint(clientX, clientY);
};
var hideThenShow = function (element, cont) {
    element.style.display = 'none';
    var result = cont();
    element.style.display = 'block';
    return result;
};
/**
 * @hidden
 */
var DraggableColumnDirective = (function () {
    function DraggableColumnDirective(draggable, element, zone, service, hint, cue, nav) {
        this.draggable = draggable;
        this.element = element;
        this.zone = zone;
        this.service = service;
        this.hint = hint;
        this.cue = cue;
        this.nav = nav;
        this.context = {};
        this.drag = new core_1.EventEmitter();
        this.subscriptions = new Subscription_1.Subscription();
    }
    Object.defineProperty(DraggableColumnDirective.prototype, "kendoDraggableColumn", {
        set: function (enabled) {
            this.enabled = enabled;
        },
        enumerable: true,
        configurable: true
    });
    DraggableColumnDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.subscriptions.add(this.zone.runOutsideAngular(function () {
            return _this.draggable.kendo.press
                .filter(function (_) { return _this.enabled; })
                .filter(function (_a) {
                var target = _a.originalEvent.target;
                return skipButtons(target);
            })
                .do(function (_a) {
                var event = _a.originalEvent;
                stopPropagation(event);
                _this.nav.navigateTo(event.target);
            })
                .switchMap(preventOnDblClick(_this.draggable.kendo.release))
                .do(function (down) {
                _this.hint.create(down, _this.element.nativeElement, _this.context.hint);
                _this.cue.create();
            })
                .switchMap(function (down) {
                return _this.draggable.kendo.drag
                    .do(_this.hint.attach())
                    .do(_this.cue.attach())
                    .takeUntil(_this.draggable.kendo.release)
                    .map(function (move) { return ({ move: move, down: down }); });
            })
                .do(_this.performDrag.bind(_this))
                .switchMapTo(_this.draggable.kendo.release)
                .subscribe(_this.drop.bind(_this));
        }));
    };
    DraggableColumnDirective.prototype.ngOnDestroy = function () {
        if (this.subscriptions) {
            this.subscriptions.unsubscribe();
        }
    };
    DraggableColumnDirective.prototype.drop = function (upEvent) {
        this.hint.remove();
        this.cue.remove();
        this.service.notifyDrop(this, upEvent);
    };
    DraggableColumnDirective.prototype.performDrag = function (_a) {
        var move = _a.move;
        this.hint.move(move);
        var cursorElement = this.elementUnderCursor(move.originalEvent);
        if (cursorElement) {
            this.service.notifyDrag(this, cursorElement, move);
        }
        this.drag.emit({
            draggable: this,
            mouseEvent: move
        });
    };
    DraggableColumnDirective.prototype.elementUnderCursor = function (mouseEvent) {
        this.hint.hide();
        var target = elementUnderCursor(mouseEvent);
        if (target && /k-grouping-dropclue/.test(target.className)) {
            target = hideThenShow(target, elementUnderCursor.bind(this, mouseEvent));
        }
        this.hint.show();
        return target;
    };
    return DraggableColumnDirective;
}());
DraggableColumnDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: '[kendoDraggableColumn]'
            },] },
];
/** @nocollapse */
DraggableColumnDirective.ctorParameters = function () { return [
    { type: draggable_directive_1.DraggableDirective, decorators: [{ type: core_1.Host },] },
    { type: core_1.ElementRef, },
    { type: core_1.NgZone, },
    { type: drag_and_drop_service_1.DragAndDropService, },
    { type: drag_hint_service_1.DragHintService, },
    { type: drop_cue_service_1.DropCueService, },
    { type: navigation_service_1.NavigationService, },
]; };
DraggableColumnDirective.propDecorators = {
    'context': [{ type: core_1.Input },],
    'kendoDraggableColumn': [{ type: core_1.Input },],
    'drag': [{ type: core_1.Output },],
};
exports.DraggableColumnDirective = DraggableColumnDirective;
